### PR TITLE
Add flexbox, third, quarter and three quarters columns, grow and stretch form fields features

### PR DESCRIPTION
**Pull request by ayumihamsaki**
## Summary of this pull request

This pr has been designed in the following logical order:

1. The main layout had to updated to flexbox to begin.
2. The top nav menu had to be fixed before I could begin on the form content area.
3. The sidebar had to be fixed before I could begin on the form content area.
4. The main content area was then corrected.
5. The form field area was then upgraded to flexbox - the rest of October CMS I left as old display:table (someone can upgrade the rest later).
6. Fixed conflicts for other areas e.g. CMS page and Settings Page had conflicts as they used the form area code. So I had to add extra code for them.
7. Small bug in builder plugin with the background when I tested all extrenal web pages for conflicts, so added a single line of CSS to this pr.
8. Can now begin to write the code to update the form fields.
9. First removed old code for all columns, then added the new columns.
10. Added the stretch and grow and merged all the columns and this feature into a single set of javascript all working as one.
11. Added role="form" to all of October CMS as I needed to get everything up to a basic standard before testing.
12. Tested each stage and passed, so now ready for reviewers to test.

---

Please don't bother to ask to split this pr up anymore; everything here all works as a single overall piece of code to add the extras columns to October. I had no choice but to update a lot of October to get things working right.

---

Below just shows things in greater detail from the above.

1. Main container is now flexbox, updated the default layout file with some new classes - github issue: https://github.com/octobercms/october/issues/4437

wwww

2. Top navigation switched to flexbox, fixed the sizes issues and right account-avatar image sizing - github issue: https://github.com/octobercms/october/issues/4005

wwww

3. CMS page sidebar, fixed the down sizing issue and now fixed to `120px` with `font-size:14px` - github issue: https://github.com/octobercms/october/issues/4005

wwww

4. CMS Page Pin button has a gap at top. On large screens there is a `2px` gap with a top nav fixed at `60px` and on smaller screens there is a `17px` gap on `45px` with a top nav fixed at 45px - github issue: https://github.com/octobercms/october/issues/4005

wwww

5. There seems to be a cross-browser error with the builder plugin, the background color is loaded in Chrome but not in firefox and so added the extra styling css `.fancy-layout.control-tabs.master-tabs > div > div.tabs-container` to october.css to bridge the gap - (new github issue discovered).

wwww

6. The various api form fields are as follows:

API field position name | Percentage | Position
--- | --- | ---
full | 100% | whole width of conatiner
left | 50% | left
right | 50% | right
auto | 50% | various
third | 33% | various
quarter | 25% | various
three-quarters | 75% | various

7. Down-scaling rules, October CMS takes care of everything with smaller screen sizes. The following rules processed:

API field position name | Media Query | Number of Fields and width values
--- | --- | ---
full | any size | 1 field (100% width)
left | above 769px |  2 fields (50% width)
left | below 769px | 1 field (100% width)
right | above 769px |  2 fields (50% width)
right | below 769px | 1 field (100% width)
auto | above 769px |  2 fields (50% width)
auto | below 769px | 1 field (100% width)
quarter | above 769px | 4 fields (25% width)
quarter | below 769px | 1 field (100% width)
third | above 769px | 3 fields (33% width)
third | below 769px | 1 field (100% width)
three-quarters | above 769px | Allows 2 different sized fields (75% width and a space for a quarter size)
three-quarters | below 769px | 1 field (100% width)

8. Fix the stretch option for the Rich editor and create two options called `stretch` and `grow`, they do the following:

- `stretch` = stretch (only works if no fields below it)

- `grow` = dynamic height (on more content)

Github issue: https://github.com/octobercms/october/issues/2303

dynamic-heights

### Outside Tabs

outside-tabs

### Primary Tabs

primary-tabs

### Secondary Tabs

secondary-tabs

9. Update builder plugin with new features, with the following new features:

### span.options

- full
- left
- right
- auto
- 3-1
- 3-2
- 3-3
- 4-1
- 4-2
- 4-3
- 4-4
- 25-left
- 25-right
- 75-left
- 75-right

### dynamic.options

- nothing
- stretch
- grow

 pr found here: https://github.com/rainlab/builder-plugin/pull/290

10. Update doc's with extra features and infomation - pr found here: https://github.com/octobercms/docs/pull/387

11. Added `role="form"` to all forms in October, helps to create a base line to add Keyboard Navigation into October CMS - github issue: https://github.com/octobercms/october/issues/4454

12. Converted to whole form field area from `display:table` into `display:flex`, now I combined my custom javascript offset code with flexbox to create complex layouts in the backend with little effect:

sssssssssss2

13. As per Luke's comment, I decided to code his request.

Stretch feature to `hint` fields:

hint

Stretch feature to `partial` fields:

partial

Grow is not added to them.

Right now it's not possible to add Stretch to YAML for these two fields, due to the code setup. Instead all you need to do is the following:

```html
<div class="stretch">
    ..
</div>
```

Add the `class stretch` to your container in your hint or partial file and October will do all the rest for you.

### CODE REVIEW

**Comment by ayumihamsaki** [comment]
A few extra points:

1. Let me know what is going on with this conflicting file message, doesn't give me any info to go on?

2. The rainlab pr is not finished and I request help to finish it, just need to inject the `stretch` and `grow` into the `class` sttribute.

3. I can fill this pr up with screenshots of what each point does to fix October if it's needed to understand this pr more. But for now I haven't because I don't want to flood these comments yet.

**Comment by ayumihamsaki** [comment]
This gif shows the `grow` and `stretch` functions (I have to manually add `grow` into the attribute due to builder pr not finished yet), but shows this pr is working:

ezgif-2-5d3b939ba0ac

**Comment by ayumihamsaki** [comment]
A few extra points:

1. Let me know what is going on with this conflicting file message, doesn't give me any info to go on?

2. The rainlab pr is not finished and I request help to finish it, just need to inject the `stretch` and `grow` into the `class` sttribute.

3. I can fill this pr up with screenshots of what each point does to fix October if it's needed to understand this pr more. But for now I haven't because I don't want to flood these comments yet.

4. Ready for testing, not going to update the commits unless a reviewer says now.

**Comment by ayumihamsaki** [comment]
### PR not finished yet - two errors still to fix:

1. October is creating incorrect YAML file codes e.g. `stretch: 1`

Need to change that to the following now:

```
dynamic: grow
dynamic: stretch
```

2. Builder plugin sidebar acting strange on flexbox, need to test more.

**Comment by ayumihamsaki** [comment]
### Last thing to fix and create

1. October is creating incorrect YAML file codes e.g. `stretch: 1`

Need to update and change to allow the following:

```
dynamic: grow
dynamic: stretch
```

**Comment by ayumihamsaki** [comment]
### Last thing to fix and create

1. October is creating incorrect YAML file codes e.g. `stretch: 1`

Need to update and change to allow the following:

```
dynamic: grow
dynamic: stretch
```

Just gonna link this here (progress): https://github.com/ayumihamsaki/backend-notes/blob/master/README.md

### Progress

Stuck on two things - any help would be great!

- nothing = working fine.

- grow = working fine.

- stretch = not working.

Not sure why not working ?

In a YAML file if you add `dynamic: stretch` it injects the stretch into the class attribute but doesn't work. If you add the `stretch: 1` into the YAML it works?

- textarea = not setup in main php file.

Not sure why not working ?

When I try to add the following line of code:

```php
$this->vars['dynamic'] = $this->formField->dynamic;
```

To code line area: https://github.com/octobercms/october/blob/217d52bc776eb19b5642ad8344955535c930f5fe/modules/backend/widgets/Form.php#L308-L316

I get an error message.

**Comment by ayumihamsaki** [comment]
### Last thing to fix and create

1. October is creating incorrect YAML file codes e.g. `stretch: 1`

Need to update and change to allow the following:

```
dynamic: grow
dynamic: stretch
```

Just gonna link this here (progress): https://github.com/ayumihamsaki/backend-notes/blob/master/README.md

---

### Progress

Stuck on one thing - any help would be great!

- nothing = working fine.

- grow = working fine.

- textarea = Need to create a stretch function for it.

- stretch = not working &larr; **stuck on this for Rich Editor**

Not sure why not working ?

In a YAML file if you add `dynamic: stretch` it injects the `stretch` into the `class` attribute but doesn't work. If you add the `stretch: 1` into the YAML it works?

https://github.com/octobercms/october/blob/c21c22e1baf1e85ceeb11999a1c1ad144f5f3a2d/modules/backend/formwidgets/richeditor/assets/js/richeditor.js#L167

https://github.com/octobercms/october/blob/c21c22e1baf1e85ceeb11999a1c1ad144f5f3a2d/modules/backend/formwidgets/richeditor/assets/js/richeditor.js#L278

**Comment by ayumihamsaki** [comment]
### Last thing to fix and create

1. October is creating incorrect YAML file codes e.g. `stretch: 1`

Need to update and change to allow the following:

```
dynamic: grow
dynamic: stretch
```

Just gonna link this here (progress): https://github.com/ayumihamsaki/backend-notes/blob/master/README.md

---

### Progress

Stuck on two things - any help would be great!

- nothing = working fine.

- grow = working fine.

- stretch = not working.

Not sure why not working ?

In a YAML file if you add `dynamic: stretch` it injects the `stretch` into the `class` attribute but doesn't work. If you add the `stretch: 1` into the YAML it works?

https://github.com/octobercms/october/blob/c21c22e1baf1e85ceeb11999a1c1ad144f5f3a2d/modules/backend/formwidgets/richeditor/assets/js/richeditor.js#L167

https://github.com/octobercms/october/blob/c21c22e1baf1e85ceeb11999a1c1ad144f5f3a2d/modules/backend/formwidgets/richeditor/assets/js/richeditor.js#L278

- textarea = not setup in main php file.

Not sure why not working ?

When I try to add the following line of code:

```php
$this->vars['dynamic'] = $this->formField->dynamic;
```

To code line area: https://github.com/octobercms/october/blob/217d52bc776eb19b5642ad8344955535c930f5fe/modules/backend/widgets/Form.php#L308-L316

I get an error message.

**Comment by ayumihamsaki** [comment]
Going to add another new feature to October just coded something to enhance textarea fields - full screen view, see gif demo:

ezgif-2-2b67fba60ff3

**Comment by ayumihamsaki** [comment]
### New feature idea

My plan is trying to make the textarea be able to do three things:

1. grow
2. stretch
3. fullscreen

Below is a code example doing the fullscreen:

ezgif-2-e00c879af276

My code repo for this update is here: https://github.com/ayumihamsaki/textarea-full-screen

**Comment by LukeTowers** [comment]
@ayumihamsaki don't make it under a single config of `dynamic`, just have it be `stretch` and `grow` as their own mutually exclusive options.

**Comment by ayumihamsaki** [comment]
WARNING: Please do not review this code yet! I plan to still do major changes to the code.

**Comment by ayumihamsaki** [comment]
@LukeTowers `Stretch` and `Grow` needs to be under a single config and not separate because you can not have this in a YAML file:

```markdown
stretch: 1
grow: 1
```

If you start growing a stretch box you would run into a lot of trouble.

**Comment by ayumihamsaki** [comment]
WARNING: Please do not review this code yet! I plan to still do major changes to the code.

**Comment by ayumihamsaki** [comment]
textarea stretch maths:

stretch

**Comment by ayumihamsaki** [comment]
Left to do:

1. Wait for Luke to reply on this comment: https://github.com/octobercms/october/issues/4437#issuecomment-513547728

2. Update doc's

3. Update builder to decided structure - span.options and remove.dynamic and do separate.

**Comment by ayumihamsaki** [comment]
## Textarea Stretch

### Outside Tabs

outside-tabs

### Primary Tabs

primary-tabs

### Secondary Tabs

secondary-tabs

**Comment by ayumihamsaki** [comment]
To do list:

1. Add the CSS for the margins for the `third`, `quarter` and `three-quarters`.

2. Update the doc's pr.

3. Update the builder pr.

**Comment by ayumihamsaki** [comment]
To do list:

1. Add the CSS for the margins for the `third`, `quarter` and `three-quarters`.

2. Update the doc's pr.

3. Update the builder pr.

4. Update doc's about:

```
{% framework %}
{% framework extras %}
{% framework extras_optimized %}
```

Also add info on Luke's comment:

> We can use preload then but in the docs we should make note of the fact that preload will load assets async in older browsers so that people are aware of the impacts of using it.

**Comment by ayumihamsaki** [comment]
To do list:

1. Add the CSS for the margins for the `third`, `quarter` and `three-quarters`.

2. Update the doc's pr.

3. Update the builder pr.

4. Update doc's about:

```
{% framework %}
{% framework extras %}
{% framework extras_optimized %}
```

Also add info on Luke's comment:

> We can use preload then but in the docs we should make note of the fact that preload will load assets async in older browsers so that people are aware of the impacts of using it.

**Comment by ayumihamsaki** [comment]
To do list:

1. Add the CSS for the margins for the `third`, `quarter` and `three-quarters`.

2. Update the doc's pr.

3. Update the builder pr.

4. Update doc's about:

```
{% framework %} 
{% framework optimized %}
{% framework extras %}
{% framework extras optimized %}
{% framework jquery %} = loads jquery sync only (not framework)
{% framework jquery optimized %} = loads jquery sync, high priorty and preload (not framework)
```

Also mention the most popular choice, that would cover most cases would be:

```
{% framework extras optimized %}
{% scripts %}
```

Also add info on Luke's comment:

> We can use preload then but in the docs we should make note of the fact that preload will load assets async in older browsers so that people are aware of the impacts of using it.

**Comment by ayumihamsaki** [comment]
To do list:

1. More Stretch to more fields, as per Luke's comment: <<< Working on now

>Stretch can apply to slightly more (`partial` & `hint` perhaps)

**Don't add grow to these two fields - stretch only.**

2. Update the doc's pr.

3. Update the builder pr.

4. Update doc's about:

```
{% framework %} 
{% framework optimized %}
{% framework extras %}
{% framework extras optimized %}
{% framework jquery %} = loads jquery sync only (not framework)
{% framework jquery optimized %} = loads jquery sync, high priorty and preload (not framework)
```

Also mention the most popular choice, that would cover most cases would be:

```
{% framework extras optimized %}
{% scripts %}
```

Also add info on Luke's comment:

> We can use preload then but in the docs we should make note of the fact that preload will load assets async in older browsers so that people are aware of the impacts of using it.

**Comment by ayumihamsaki** [comment]
To do list:

1. jQuery ready()

2. More Stretch to more fields, as per Luke's comment:

>Stretch can apply to slightly more (`partial` & `hint` perhaps)

**Don't add grow to these two fields - stretch only.**

3. Update the doc's pr.

4. Update the builder pr.

5. Update doc's about:

```
{% framework %} 
{% framework optimized %}
{% framework extras %}
{% framework extras optimized %}
{% framework jquery %} = loads jquery sync only (not framework)
{% framework jquery optimized %} = loads jquery sync, high priorty and preload (not framework)
```

Also mention the most popular choice, that would cover most cases would be:

```
{% framework extras optimized %}
{% scripts %}
```

Also add info on Luke's comment:

> We can use preload then but in the docs we should make note of the fact that preload will load assets async in older browsers so that people are aware of the impacts of using it.

**Comment by ayumihamsaki** [comment]
To do list:

1. Grid layout two more things to do:

- Variable margin for non-standard layouts.
- Nothing to the right, snap to right hand side.

2. Update the builder pr (wait for Luke reply on a review).

3. Update doc's about:

```
{% framework %} 
{% framework optimized %}
{% framework extras %}
{% framework extras optimized %}
{% framework jquery %} = loads jquery sync only (not framework)
{% framework jquery optimized %} = loads jquery sync, high priorty and preload (not framework)
```

Also mention the most popular choice, that would cover most cases would be:

```
{% framework extras optimized %}
{% scripts %}
```

Also add info on Luke's comment:

> We can use preload then but in the docs we should make note of the fact that preload will load assets async in older browsers so that people are aware of the impacts of using it.

**Comment by ayumihamsaki** [comment]
## To do list

1. Grid layout two more things to do:

- Variable margin for non-standard layouts.
- Nothing to the right, snap to right hand side.

2. Update the builder pr (wait for Luke reply on a review).

3. Maybe doc's are finsihed now - wait for reviewers.

**Comment by ayumihamsaki** [comment]
Testing layout:

ssss

File: modules/backend/assets/js/october.sidenav-tree.js [hunk 1]
```diff
@@ -25,7 +25,7 @@
     }
 
     SidenavTree.DEFAULTS = {
-        treeName: 'sidenav_tree'
+        treeName: 'sidenav_tree'r
```
Comment (FlusherDock1) [inline modules/backend/assets/js/october.sidenav-tree.js]: Found typo here

Comment (ayumihamsaki) [reply modules/backend/assets/js/october.sidenav-tree.js]: Nice one, thanks for the info! 👍

**Comment by ayumihamsaki** [comment]
## To do list

1. Grid layout two more things to do: <<< working on still.

- Variable margin for non-standard layouts - done.
- Nothing to the right, snap to right hand side - done.

Finish off the flexbox updates...

- Roll it out for the Primary and Secondary tabs.
- Test all the complex widgets e.g. Rich Editor etc.

Add 18. Flexbox firld form area to above summary when finished and passed testing.

2. Update the builder pr (Luke said to switch the whitelist and blacklist around for all the checkboxes).

3. Maybe doc's are finsihed now - wait for reviewers.

**Comment by ayumihamsaki** [comment]
## To do list

1. Grid layout two more things to do: <<< working on still.

- js offset needs to update when a user opens a new tab - event listener.

Update summary: 18. === Re-Code Builder plugin to use Flexbox and add the new drag n drop columns ===

2. Update the builder pr

- Luke said to switch the whitelist and blacklist around for all the checkboxes.
- Rip out the old css and convert builder css into flexbox as well.

3. Maybe doc's are finsihed now - wait for reviewers.

**Comment by ayumihamsaki** [comment]
## To do list

1. Grid layout two more things to do: <<< working on still.

- Form Area Tabs

- [x] - Outside Tabs
- [x] - Primary Tabs
- [ ] - Secondary Tabs
- [ ] - Master Tabs
- [ ] - Content Tabs

2. Update the builder pr

- Luke said to switch the whitelist and blacklist around for all the checkboxes.
- Rip out the old css and convert builder css into flexbox as well.

3. Maybe doc's are finsihed now - wait for reviewers.

**Comment by ayumihamsaki** [comment]
## To do list

1. Grid layout two more things to do:

- Form Area Tabs

- [x] - Outside Tabs
- [x] - Primary Tabs
- [x] - Secondary Tabs
- [x] - Master Tabs
- [x] - Content Tabs

2. Update the builder pr <<< working on still.

- Luke said to switch the whitelist and blacklist around for all the checkboxes.
- Rip out the old css and convert builder css into flexbox as well.

3. Maybe doc's are finsihed now - wait for reviewers.

**Comment by ayumihamsaki** [comment]
## To do list

1. Last thing, need to find out why this is happening now:

last-issue

2. Update the builder pr

- Luke said to switch the whitelist and blacklist around for all the checkboxes.
- Rip out the old css and convert builder css into flexbox as well.

3. Maybe doc's are finsihed now - wait for reviewers.

**Comment by ayumihamsaki** [comment]
## To do list

1. Last thing, need to sort out:

last-issue

2. Update the builder pr

- Luke said to switch the whitelist and blacklist around for all the checkboxes.
- Rip out the old css and convert builder css into flexbox as well.

3. Maybe doc's are finsihed now - wait for reviewers.

**Comment by ayumihamsaki** [comment]
## IMPORTANT MESSAGE

This message is here to try and prevent people in the future from changing the code in an important section.

The styling for the tabs has been preloaded by displaying the hidden tabs and the javascript calculates the widths and positions of all the tabs. This is all done when a user loads the web page for the first time. By the time the user opens one of the hidden tab panels, all the fields layouts will have been calculated. This gives a good performance for the user! This is done by using `opacity:0` and `height:0`. If someone sets it to display:none you will break October. The active tab is displayed using `opacity:1` and `height:100%` and `display:flex`. We are not using `display:table` anymore that has all been romved. This section of code is important and this notice is intended for future users to not change the code in this section.

All of this can be found in the commit: https://github.com/octobercms/october/pull/4463/commits/0f43155fb805cd558a793aca0c61350f450ae3dc

**Comment by ayumihamsaki** [comment]
Test results:

localhost_json_ayumi_tabs_index_update_1

**Comment by ayumihamsaki** [comment]
@LukeTowers @w20k @bennothommo @FlusherDock1 

This PR is now finished and ready for testing.

You can find a random YAML models test file here: https://github.com/ayumihamsaki/yaml-test-file/blob/master/fields.yaml

**Comment by ayumihamsaki** [comment]
> Conflicting files
> modules/backend/layouts/default.htm

This is a change to my previous merged pr, the commit file is here: https://github.com/octobercms/october/blob/0666e959e6f97ed9a22a33904c4a4cc2affd9fa3/modules/backend/layouts/default.htm

Please accept this conflict update in this pr.

**Comment by ayumihamsaki** [comment]
> Conflicting files
> modules/backend/layouts/default.htm

This is a change to my previous merged pr, the commit file is here: https://github.com/octobercms/october/blob/0666e959e6f97ed9a22a33904c4a4cc2affd9fa3/modules/backend/layouts/default.htm

Please accept this conflict update in this pr.

Other PR (merged) has the following code:

```html
<div class="control-breadcrumb" aria-label="<?= e(trans('system::lang.aria-label.breadcrumb')) ?>">
    <?= $breadcrumbContent ?>
</div>
```

This PR corrects the code into this:

```html
<div class="control-breadcrumb">
    <nav aria-label="<?= e(trans('system::lang.aria-label.breadcrumb')) ?>">
        <?= $breadcrumbContent ?>
    </nav>
</div>
```

I guess when you update the stable verision, this pr will switch to green and have no conflicting notice.

**Comment by ayumihamsaki** [comment]
Finished.

**Comment by ayumihamsaki** [comment]
Finished.

**Comment by ayumihamsaki** [comment]
Issue with Travis while trying to rebase this pr, closed this pr and moved to: https://github.com/octobercms/october/pull/4496

**Comment by ayumihamsaki** [comment]
My dev branch is all screwed up, going to wait for all the pr's to get merged and I will delete my fork and do everything fresh - I will create a new for this one.

**Pull request closed without merging**